### ISSUE TITLE
-D<PROJECT_NAME>_SCALE_TEST_TIMEOUT does not get applied to default timeout

### ISSUE THREAD

**Issue by bkochuna** [TriBITSPub/TriBITS#164]
We have a CMake function like the following

```
FUNCTION(CreateUnitTest TESTNAME)
    TRIBITS_ADD_EXECUTABLE_AND_TEST(${TESTNAME}
        SOURCES ${TESTNAME}.f90
        NUM_MPI_PROCS 1
        LINKER_LANGUAGE Fortran
    )
ENDFUNCTION()
```
and for our test configuration it appears that TriBITS determines the test timeout  to be 600 s (in other configurations it appears that TriBITS will determine the timeout to be 1500 s).

It's not clear how the default timeout for the test is determined.

Furthermore, if use the `-D<PROJECT_NAME>_SCALE_TEST_TIMEOUT` option, this does not affect the test timeout determined by TriBITS for a test defined like in the example above.

To resolve this issue, the following tasks are suggested:
1. the documentation be updated to provide some detail on how the default timeout for a test is determined
2. that the  `-D<PROJECT_NAME>_SCALE_TEST_TIMEOUT` value be used to scale the default test timeout

**Comment by bartlettroscoe** [comment]
Thanks @bkochuna, I will have a look.  I will start with trying to make the documentation more clear how test timeouts are determined and then see if `TRIBITS_ADD_EXECUTABLE_AND_TEST()` is doing this correctly or not.

**Comment by bkochuna** [comment]
Hi Ross,

Is there any update on this? Or do you have a timeline on when this might be ready?

**Comment by bartlettroscoe** [comment]
@bkochuna, to see if this is a defect with `TRIBITS_ADD_EXECUTABLE_AND_TEST()`, what happens when you split this up into two separate calls `TRIBITS_ADD_EXECUTABLE()` and `TRIBITS_ADD__TEST()`?  I am looking at the CMake source code for the `TRIBITS_ADD_EXECUTABLE_AND_TEST()` and it looks like it does not pass in any argument for `TIMEOUT` to `TRIBITS_ADD__TEST()` which means means it should just use the default timeout.  I will check over the unit tests and write up how timeouts are handled a little more explicitly with the documentation.

**Comment by bartlettroscoe** [comment]
@bkochuna, I added a new section:

* https://tribits.org/doc/TribitsDevelopersGuide.html#setting-timeouts-for-tests-tribits-add-test

that should hopefully explain how timeouts are handled with CMake, CTest, and TriBITS and how CTest deals with timeouts in the end (to help debug things).

I tested this manually with the TriBITS project itself and it does seem to be broken:

```
$ ./do-configure -DDART_TESTING_TIMEOUT=2 -DTriBITS_SCALE_TEST_TIMEOUT=3

$ grep DART_TESTING_TIMEOUT CMakeCache.txt  | grep "="
DART_TESTING_TIMEOUT:STRING=2
DART_TESTING_TIMEOUT-ADVANCED:INTERNAL=1

$ grep TimeOut DartConfiguration.tcl | grep ":"
TimeOut: 2
```

'TimeOut' should be '6' not '2'.

I will dig into this (and why my [automated unit test](http://my.cdash.org/testDetails.php?test=29082656&build=1103741) does not sufficiently protect this).

**Comment by bkochuna** [comment]
Thanks for the updated documentation Ross. It is very clear now on what I can check to debug in the future. Puzzling though that the SCALE_TEST_TIMEOUT isn't being applied.

**Comment by bartlettroscoe** [comment]
I have added a unit test that exposes the problem.  The issue is that the first time through, CMake is not writing the updated value of `DART_TESTING_TIMEOUT` as 'TimeOut' into the DartConfiguration.tcl file.

The configure shows that it should be getting scaled as:

```
$ ./do-configure -DDART_TESTING_TIMEOUT=2 -DTriBITS_SCALE_TEST_TIMEOUT=3
...
Setting up testing support ...

-- DART_TESTING_TIMEOUT=2 being scaled by TriBITS_SCALE_TEST_TIMEOUT=3 to 6
...
```

But the DartConfiguration.tcl has the unscaled value:

```
$ grep TimeOut: DartConfiguration.tcl 
TimeOut: 2
```

But if you must reconfigure, it is set correctly:

```
$ cmake .

$ grep TimeOut: DartConfiguration.tcl 
TimeOut: 6
```

That is why the existing unit tests failed.  They only tested scaling on reconfigures (to make sure that reconfiguring over and over would not keep scaling the timeouts up an up).

Now to figure out why CMake is not setting the correct value the first time out.  Every indication is that  `DART_TESTING_TIMEOUT=6` is set correctly before the INCLUDE(CTest) statement that generates the DartConfiguration.tcl file.

**Comment by bartlettroscoe** [comment]
@bkochuna, I have got a fix pushed to TriBITS GitHub 'master' in the commit 3370b5e.  Where all do you need this fixed?  Note that I can't just merge TriBITS GitHub 'master' to the CASL TriBITS casl-dev 'master' branch because Trilinos would need to be updated first.  See:

* https://vminfo.casl.gov/trac/casl_phi_kanban/ticket/3913#comment:12

So if you need to have this fixed in CASL TriBITS casl-dev 'master' branch, I will need to cherry-pick this commit.

Let me know,

-Ross

**Comment by bkochuna** [comment]
Hi Ross,

We specifically need this for the  UM CI, so its possible we could snapshot TriBITS into MPACT, but then we'd have different dev env's for CASL and UM which might cause a different set of issues. It might be best to cherry-pick that commit if it is not too much trouble. Thanks for the rapid turn around on investigating and fixing!

**Comment by bartlettroscoe** [comment]
Okay, I will cherry-pick the commit 3370b5e to the TriBITS casl-dev master branch and push.  I will let you know when it is done.  Then you can just update the snapshot of TriBITS in MPACT from the casl-dev master branch.

**Comment by bartlettroscoe** [comment]
@bkochuna,

I just cherry-picked this commit and pushed it to the 'master' branch at `git@casl-dev:TriBITS` as the commit:

```
939827a "Fix <Project>_SCALE_TEST_TIMEOUT on first config (#164)"
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Thu Dec 22 14:03:31 2016 -0700 (30 minutes ago)

M       test/core/ExamplesUnitTests/CMakeLists.txt
M       tribits/core/package_arch/TribitsGlobalMacros.cmake
```

Please try out that version of TriBITS with MPACT and see if it fixes the error you are seeing.  In any case, please report your results back here so that we can fix the problem or close the issue as being fixed.

Note that this defect never really affected the scaling of timeouts for Trilinos or CASL VERA that used TribitsCTestDriverCore.cmake and the first package configured, built, and tested did not need scaling.  For CASL VERA, the first package is TriBITS which does not need timeouts scaling on most platforms (since all of the tests run very fast compared to most tests in downstream VERA packages).  That is why we never noticed this defect before.

**Comment by bartlettroscoe** [comment]
@bkochuna, any chance to try this yet?  Is it working for you?

**Comment by bkochuna** [comment]
Hi Ross,

I did get a chance to try this today. The only thing I noticed is that `-D<Project>_SCALE_TEST_TIMEOUT` only seems to work if `-DDART_TESTING_TIMEOUT` is defined.

If `-DDART_TESTING_TIMEOUT` is not defined when the cmake configure is invoked it appears that
`-D<Project>_SCALE_TEST_TIMEOUT` does nothing.

For our project the `DART_TESTING_TIMEOUT` is eventually determined to be 1500 s.

We can probably make this work as is, but I'm wondering if you want to have
`-D<Project>_SCALE_TEST_TIMEOUT` still work when `-DDART_TESTING_TIMEOUT` is undefined.

I looked at the documentation here: https://tribits.org/doc/TribitsBuildReference.html#dart-testing-timeout

I noticed two things about this part of the documentation
1. It does not say how this value is set if it is not defined.
2. It refers to `<Project>_SCALE_TEST_TIMEOUT_TESTING_TIMEOUT` is this a typo?

Do you want to address these things as a part of this issue or would you like me to open a new issue?

**Comment by bartlettroscoe** [comment]
@bkochuna,

> The only thing I noticed is that `-D<Project>_SCALE_TEST_TIMEOUT` only seems to work if `-DDART_TESTING_TIMEOUT` is defined.  If `-DDART_TESTING_TIMEOUT` is not defined when the cmake configure is invoked it appears that `-D<Project>_SCALE_TEST_TIMEOUT` does nothing.

That is correct.  For tests that don't have a customized `TIMEOUT` property set (i.e. uses the global default timeout), and `DART_TESTING_TIMEOUT` is not set (i.e. is implicitly infinity) then setting `<Project>_SCALE_TEST_TIMEOUT` does nothing (because scaling infinity by a finite number is still infinity) to affect the timeout of such tests.  But the `ctest` option `--ctest-timeout` would still set a timeout for these tests but `<Project>_SCALE_TEST_TIMEOUT` would not impact that (because CTest does not know about this cache var).

> For our project the `DART_TESTING_TIMEOUT` is eventually determined to be 1500 s.

How does that happen if you don't set `DART_TESTING_TIMEOUT`?

> We can probably make this work as is, but I'm wondering if you want to have
`-D<Project>_SCALE_TEST_TIMEOUT` still work when `-DDART_TESTING_TIMEOUT` is undefined.

In that case it would only scale individual test timeouts specified using the `TIMEOUT` argument.  It can't scale a default global timeout since none is defined.

> I looked at the documentation here: https://tribits.org/doc/TribitsBuildReference.html#dart-testing-timeout
>
> I noticed two things about this part of the documentation
> 1. It does not say how this value is set if it is not defined.

Which value?  Do you mean  `DART_TESTING_TIMEOUT` or `<Project>_SCALE_TEST_TIMEOUT`?  By default, both are undefined and are therefore implicitly infinity and 1.0, respectively.  Does that make sense?

> 2. It refers to `<Project>_SCALE_TEST_TIMEOUT_TESTING_TIMEOUT` is this a typo?

Yup.  That was typo.  This should be fixed now.

> Do you want to address these things as a part of this issue or would you like me to open a new issue?

No, we can address this in this issue.

Please have a look at the updated cross-linked sections:

* https://tribits.org/doc/TribitsDevelopersGuide.html#setting-timeouts-for-tests-tribits-add-test (timeouts from the developer's perspective)
* https://tribits.org/doc/TribitsBuildReference.html#dart-testing-timeout (from the user perspective)
* https://tribits.org/doc/TribitsBuildReference.html#project-scale-test-timeout-testing-timeout (from the user perspective)

Please let me know if that makes sense or not.  Otherwise, let's try to pin down the point of confusion and how to improve the documentation to address this.

**Comment by bkochuna** [comment]
Ross, thanks again for the enlightening reply. Your explanation makes very good sense. A few responses to your questions below.

> How does that happen if you don't set DART_TESTING_TIMEOUT?

This is a very good question. I'll see what I can figure out examining our project. This is likely the biggest point of confusion for me.

> Which value? Do you mean DART_TESTING_TIMEOUT or <Project>_SCALE_TEST_TIMEOUT? By default, both are undefined and are therefore implicitly infinity and 1.0, respectively. Does that make sense?

Yes, I meant DART_TESTING_TIMEOUT. The explanation makes perfect sense, perhaps just update the documentation to say what the values are if they are undefined. (And maybe you did this, I have not looked back at the documentation)

**Comment by bartlettroscoe** [comment]
> The explanation makes perfect sense, perhaps just update the documentation to say what the values are if they are undefined.

I updated the documentation at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#setting-timeouts-for-tests-tribits-add-test
* https://tribits.org/doc/TribitsBuildReference.html#dart-testing-timeout

to try to make this more explicit.  Please have a look if you have a few minutes.

I am thinking that it might be good for projects to be able to define a default global test timeout by setting a variable called `DART_TESTING_TIMEOUT_DEFAULT` set in a proejct's base CMakeLists.txt file (or its ProjectName.cmake file).  The user could override this by setting `DART_TESTING_TIMEOUT` in the CMake cache.  But if they did not, then at least every test would get some default timeout for every build.  This would get documented with other project-level settings at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#tribits-global-project-settings

Would  `DART_TESTING_TIMEOUT_DEFAULT` be helpful?

**Comment by bkochuna** [comment]
> Would `DART_TESTING_TIMEOUT_DEFAULT` be helpful?
Maybe... However, if we leave it undefined, then I think we are in the same boat that we are now. Its not obvious to me that there is a clear value added with that if that's the case.

Documentation looks great!

I can't find any reference to the 1500 seconds in our project by grepping the source.

Do you have any other suggestions on how to go about debugging where that value could be getting set?

My next attempt was going to start putting print statements in CMake/Tribits to echo the value of `DART_TESTING_TIMEOUT` and try to find where it gets set.

**Comment by bartlettroscoe** [comment]
> I can't find any reference to the 1500 seconds in our project by grepping the source.

What, specifically, is getting set to `1500`?  What value in what output file?  What do you see for `DART_TESTING_TIMOUT` in your CMakeCache.txt file?  What do you see for your `TimeOut` property in your DartConfiguration.tcl file?  What value for `TIMEOUT` for the tests in question are you seeing in the CTestTestfile.cmake files?

**Comment by bkochuna** [comment]
`DART_TESTING_TIMEOUT` is being set to 1500 in the CMakeCache.txt file.
`TimeOut` in the DartConfiguration.tcl is 1500.
The CTestTestfile.cmake in the  build directory does not have TIMEOUT.

**Comment by bartlettroscoe** [comment]
You should be able to debug this with a few MESSAGE() print statements in CMake code.  You may need to modify your local version of TriBITS to put in print statements.  You can see where TriBITS deals with `DART_TESTING_TIMEOUT` by looking at:

```
[rabartl@crf450 tribits (master)]$ find . -name "*.cmake" -exec grep -nH DART_TESTING_TIMEOUT {} \;
./core/package_arch/TribitsGlobalMacros.cmake:493:    "Scale factor for global DART_TESTING_TIMEOUT and individual test TIMEOUT (default 1.0)."
./core/package_arch/TribitsGlobalMacros.cmake:709:  MARK_AS_ADVANCED(DART_TESTING_TIMEOUT)
./core/package_arch/TribitsGlobalMacros.cmake:1930:  SET(DART_TESTING_TIMEOUT_IN ${DART_TESTING_TIMEOUT})
./core/package_arch/TribitsGlobalMacros.cmake:1932:  IF (DART_TESTING_TIMEOUT_IN)
./core/package_arch/TribitsGlobalMacros.cmake:1933:    TRIBITS_SCALE_TIMEOUT(${DART_TESTING_TIMEOUT} DART_TESTING_TIMEOUT)
./core/package_arch/TribitsGlobalMacros.cmake:1934:    IF (NOT DART_TESTING_TIMEOUT STREQUAL DART_TESTING_TIMEOUT_IN)
./core/package_arch/TribitsGlobalMacros.cmake:1935:     MESSAGE("-- DART_TESTING_TIMEOUT=${DART_TESTING_TIMEOUT_IN} being scaled by ${PROJECT_NAME}_SCALE_TEST_TIMEOUT=${${PROJECT_NAME}_SCALE_TEST_TIMEOUT} to ${DART_TESTING_TIMEOUT}")
./core/package_arch/TribitsGlobalMacros.cmake:1937:    # Have to set DART_TESTING_TIMEOUT in cache or CMake will not put in right
./core/package_arch/TribitsGlobalMacros.cmake:1939:    SET(DART_TESTING_TIMEOUT ${DART_TESTING_TIMEOUT} CACHE STRING "" FORCE)
./core/package_arch/TribitsGlobalMacros.cmake:1944:  IF (DART_TESTING_TIMEOUT_IN)
./core/package_arch/TribitsGlobalMacros.cmake:1945:    # Put DART_TESTING_TIMEOUT back to user input value to avoid scaling this
./core/package_arch/TribitsGlobalMacros.cmake:1947:    SET(DART_TESTING_TIMEOUT ${DART_TESTING_TIMEOUT_IN} CACHE STRING
./core/package_arch/TribitsGlobalMacros.cmake:1955:# NOTE: The above logic with DART_TESTING_TIMEOUT is a huge hack.  For some
./core/package_arch/TribitsGlobalMacros.cmake:1957:# scaled DART_TESTING_TIMEOUT variable into the DartConfiguration.tcl.
./core/package_arch/TribitsGlobalMacros.cmake:1958:# Instead, it uses the value of DART_TESTING_TIMEOUT that is in the cache.
./core/package_arch/TribitsGlobalMacros.cmake:1960:# DART_TESTING_TIMEOUT and ignores the value in the cache (very irritating).
./core/package_arch/TribitsGlobalMacros.cmake:1962:# force-set DART_TESTING_TIMEOUT in the cache on the first configure.  But to
./core/package_arch/TribitsGlobalMacros.cmake:1963:# avoid rescaling DART_TESTING_TIMEOUT up and up on reconfigures, you have to
./core/package_arch/TribitsGlobalMacros.cmake:1964:# force-set DART_TESTING_TIMEOUT back to the user's input value.  The only
```

With CMake, it is hard to see when a variable gets set.  You basically have to resort to well-placed print statements (but that is a pretty fast debugging process).

**Comment by bkochuna** [comment]
Found it. In `./core/package_arch/TribitsGlobalMacros.cmake` you have:

```
1864 MACRO(TRIBITS_INCLUDE_CTEST_SUPPORT)
1865   IF (DART_TESTING_TIMEOUT)
1866     SET(DART_TESTING_TIMEOUT_IN ${DART_TESTING_TIMEOUT})
1867     TRIBITS_SCALE_TIMEOUT(${DART_TESTING_TIMEOUT} DART_TESTING_TIMEOUT)
1868     IF (NOT DART_TESTING_TIMEOUT STREQUAL DART_TESTING_TIMEOUT_IN)
1869      MESSAGE("-- DART_TESTING_TIMEOUT=${DART_TESTING_TIMEOUT_IN} being scaled by ${PROJECT_NAME}_SCALE_TEST_TIMEOUT=${${PROJECT_NAME}_SCALE_TEST_TIMEOUT} to ${DART_TESTING_TIMEOUT}")
1870     ENDIF()
1871   ENDIF()
1872   INCLUDE(CTest)
1873   TRIBITS_CONFIGURE_CTEST_CUSTOM(${${PROJECT_NAME}_SOURCE_DIR}
1874     ${${PROJECT_NAME}_BINARY_DIR})
1875 ENDMACRO()
```

At line 1872 the call to `INCLUDE(CTest)` resolves to the CTest.cmake file located in the `<CMAKE_INSTALL_DIR>/share/cmake-3.3/Modules/CTest.cmake` (at least for version 3.3.x).

For version 3.3.2 we have in this file:
```
173   set(DART_TESTING_TIMEOUT 1500 CACHE STRING
174    "Maximum time allowed before CTest will kill the test.")
```

I'll also note, that if I run `ctest -VV` for a particular unit test from the top level build directory ctest reports a timeout of 1500s to the prompt. However, if I run that same test from deeper in the build directory ctest reports a timeout of 9.99988e+06 (probably infinity).

Does it make sense to use the value for the `DART_TESTING_TIMEOUT` defined by the CMake/CTest install as the default and have the SCALE_TEST_TIMEOUT operate on this?

Do we know when that got added to CMake? Is that backward compatible to the earliest version of CMake supported by TriBITS?

Perhaps the ` MACRO(TRIBITS_INCLUDE_CTEST_SUPPORT)` routine can be changed?
(I'm not quite sure how to do this).

**Comment by bartlettroscoe** [comment]
> `<CMAKE_INSTALL_DIR>/share/cmake-3.3/Modules/CTest.cmake` (at least for version 3.3.x).
>
> For version 3.3.2 we have in this file:
> 
> ```
> 173   set(DART_TESTING_TIMEOUT 1500 CACHE STRING
> 174    "Maximum time allowed before CTest will kill the test.")
> ```

Wow, I did not realize that CMake started adding a default value for DART_TESTING_TIMEOUT.   I have just always set this in every automated build so I never would have noticed when that default got added. We could search the Git repo for CMake and see when they did this but that will not really help us since we can't scale the value if it is defined inside of CTest.cmake and then written into the DartConfiguration.tcl file.

I think the way to address this is to have TriBITS define the variable DART_TESTING_TIMEOUT_DEFAULT, give it the default value of 1500, and then set the cache var DART_TESTING_TIMEOUT default to that.  In other words, TriBITS would do:

```
IF ( "${DART_TESTING_TIMEOUT_DEFAULT}"  STREQUAL  "")
  SET(DART_TESTING_TIMEOUT_DEFAULT 1500)
ENDIF()
SET( DART_TESTING_TIMEOUT  ${DART_TESTING_TIMEOUT_DEFAULT}
  CACHE STRING "Blah blah blah ...")
```

That way, it would have the same default that CMake provides (at least in newer versions) but it would scale it correctly by `<Project>_SCALE_TEST_TIMEOUT`.

Of course then we would update the documentation for this update.

Does that sounds like the right thing to do?

**Comment by bkochuna** [comment]
> Does that sounds like the right thing to do?
Yes. In the mean time I may just go ahead and define the DART_TESTING_TIMEOUT variable in MPACT's top level CMakeLists.txt file.

So there won't be any rush (from our perspective) to get this completed and cherry-picked to casl-dev.

But I would say once that mod is made and the documentation is updated this issue can be closed. I can review the documentation again.

**Comment by bartlettroscoe** [comment]
To avoid a cherry-pick and get this working correctly quickly, you can just explicitly set the default for `DART_TESTING_TIMEOUT` in MPACT/CMakeLists.txt and you should be good to go.  I just want to make sure 

But given given that the updated documentation is actually not correct for more recent versions of CMake, I will go ahead and make the change described [above](https://github.com/TriBITSPub/TriBITS/issues/164#issuecomment-271702018).  That is, what I will do is:

* Define `DART_TESTING_TIMEOUT_DEFAULT`, give it a default value of 1500 and set the cache var `DART_TESTING_TIMEOUT` as described [above](https://github.com/TriBITSPub/TriBITS/issues/164#issuecomment-271702018).

* Update the TriBITS automated tests to test this use case and confirm that the default timeout is 1500 and if one only explicitly sets the scaling `<Project>_SCALE_TEST_TIMEOUT` then it will scale that default value.

* Update the documentation to note the default value of `DART_TESTING_TIMEOUT` and the scaling of that value by `<Project>_SCALE_TEST_TIMEOUT`.

**Comment by bartlettroscoe** [comment]
Boy, there is just no offical CMake documenation that I can find that says that DART_TESTING_TIMEOUT has a default of 1500.  All that I can find (other than digging into CTest.cmake) is:

* http://stackoverflow.com/questions/3545598/using-cmake-with-ctest-and-cdash
* https://cmake.org/pipermail/cmake-developers/2015-December/027278.html

Very frustrating.  I feel like so much TriBITS documentation is just documenting and explaining raw CMake.

**Comment by bartlettroscoe** [comment]
@bkochuna,

I updated the implementation and documentation as described [above](https://github.com/TriBITSPub/TriBITS/issues/164#issuecomment-271913413) and pushed to 'master' in this repo.  The updated documentation can be seen at:

* https://tribits.org/doc/TribitsBuildReference.html#dart-testing-timeout
* https://tribits.org/doc/TribitsBuildReference.html#project-scale-test-timeout
* https://tribits.org/doc/TribitsDevelopersGuide.html#dart-testing-timeout
* https://tribits.org/doc/TribitsDevelopersGuide.html#setting-timeouts-for-tests-tribits-add-test

Hopefully this is now all correct.  I put in some pretty good automated tests that assert the correct behavior.  That is, there is now a test that makes sure that it will scale the default timeout of 1500 seconds right in the first configure.  See

* http://my.cdash.org/testSummary.php?project=902&name=TriBITS_TribitsHelloWorld_DefaultGlobalTimeout_ScaleTimeout&date=2017-01-11

However, again, I can't sync this to the CASL VERA copy of TriBITS until Trilinos gets updated (or I fix the old CASL copy of Trilinos for the updated TriBITS).

Please let me know if setting:

```
SET(DART_TESTING_TIMEOUT 1500 CACHE STRING "...")
```

in MPACT/CMakeLists.txt does not do the trick.  (NOTE: you have to set this as a cache var or Modules/CTest.cmake will not deal with it correctly).

Otherwise, please let me know if we can close this Issue ticket.

Thanks,

-Ross

**Comment by bkochuna** [comment]
Ross, the documentation looks great.

I added  `SET(DART_TESTING_TIMEOUT 1500 CACHE STRING "...")` to our project's top level CMakeLists.txt file.

I checked our CDash log from last night for our valgrind builds, and indeed, I observe the following in the configure output:
```
Setting up testing support ...

-- DART_TESTING_TIMEOUT=1500 being scaled by MPACT_SCALE_TEST_TIMEOUT=4000.0 to 6000000
```

I also verified that in the build directory for this CI test configuration that
```
countm@sunspear:~/CI/SERIAL_DEBUG_CDASH_EXTRAS/BUILD$ grep "DART_TESTING_TIMEOUT" CMakeCache.txt
DART_TESTING_TIMEOUT:STRING=1500
//ADVANCED property for variable: DART_TESTING_TIMEOUT
DART_TESTING_TIMEOUT-ADVANCED:INTERNAL=1
```
```
$ grep "TimeOut:" DartConfiguration.tcl
TimeOut: 6000000
```

However... it looks like the same tests still failed due to timing out when running under valgrind...

I'll try wiping the build directory and rerunning this CI test configuration later today. I'm not sure why they still would have timed out.

**Comment by bartlettroscoe** [comment]
@bkochuna,

Does the timeout on your valgrind test say `6000000` or `1500`?  Does CDash not tell you that?  Hum, I guess Kitware never fixed that:

* https://docs.google.com/document/d/1TLHRp8eTNKw7udOhwIxrOYShXQUbxAzsXeOq5cwWnKM/edit#bookmark=id.6vjbti5h6swv

But from that backlog item description, it seems you can see this from the ‘ctest -V -S’ output from the test machine itself (if you saved that output).  Very irritating.

The backlog for missing CDash features is just huge.  This is just one example.

**Comment by bartlettroscoe** [comment]
> CDash does not tell us the timeout for the valgrind.

But if you run `ctest -T memcheck -R <test_name>` locally as described here:

* https://tribits.org/doc/TribitsBuildReference.html#running-memory-checking

then the `ctest` output will show you the runtime for the test when run with valgrind.  You can also see it in the `ctest -V -S <script>` output (if you have saved and archived that).

> I think we've fixed the problem, so I'm not sure what else to do here in these issues.
>
> This issue is good to close as far as I'm concerned.

Thanks!  Closing as complete.

**Comment by bkochuna** [comment]
Thanks for the last comment Ross. I thought there was a way to do this, but wasn't sure what it was. I did
```
$ ctest -VV -T memcheck -R <test_name>
```
on one of the tests that we assume times out and prints the correct scaled timeout to the screen (`Test timeout computed to be: 6e+06`). 

Is there a way I can found what CMake executes for the test when the CI script runs?

**Comment by bartlettroscoe** [comment]
> Is there a way I can found what CMake executes for the test when the CI script runs?

What do you mean by "what CMake executes for the test"?  The timeout that it computes?  What it is running?

**Comment by bkochuna** [comment]
So we use cron to launch a test driver script that ultimately calls a function in TriBITS, and CMake runs all the commands to configure, build, and test over multiple configurations for package enables/disables.

My question is, can I look at one of these logs (or CDash) to see exactly what command got executed to that ran a particular test.

e.g. Did the CI execute the command `ctest -VV -T memcheck -R <test_name>` or did it execute something else? If so, what was it?

(hopefully that is clearer?)

**Comment by bartlettroscoe** [comment]
> My question is, can I look at one of these logs (or CDash) to see exactly what command got executed to that ran a particular test.

Here is my advice ...

Use `ctest -V -S <script>` to run your CTest driver script and archive the output so that you can look at it while it is running and when it is finished (for the next day or so then write over it).  This is how I run the current Trilinos CI server that I am running:

```
cd /ascldap/users/rabartl/Trilinos.base/SEMSCIBuild && cp trilinos_ci_server.out trilinos_ci_server.last.out && ./Trilinos/cmake/ctest/drivers/sems_ci/trilinos_ci_sever.sh &> trilinos_ci_server.out
```
That way I can watch the output while it is running and see the output from the prior day.  (NOTE: I did **NOT** use the TriBITS Dashboard Driver system to posts to CDash because I hate that system and CTest will not show you the output from a test that it is running while it is running.)

The `ctest -V -S <script>` output shows you what the runtimes are for the valgrind tests.  You can't see that anywhere on CDash currently (as mentioned above).

Every test on CDash shows you the exactly that it used to run the test.  For example, the TriBITS test `TriBITS_CTestScripts_TAAT_will_fail_pass_regular_expression_fail_regular_expression` with its last output here:

* http://my.cdash.org/testDetails.php?test=28210190&build=1115704

shows the command used to run the test at the top of the page under "Show Command Line" which shows:

```
/usr/bin/cmake "-DTEST_CONFIG=" "-P" "/home/travis/build/TriBITSPub/TriBITS/build/BUILD/test/core/CTestScriptsUnitTests/TriBITS_CTestScripts_TAAT_will_fail_pass_regular_expression_fail_regular_expression.cmake"
```

And then, of course, the tests added with TRIBITS_ADD_ADVANCED_TESTS() show you every command run and its output:

```
================================================================================

TEST_0

Running: "/home/travis/build/TriBITSPub/TriBITS/build/BUILD/test/core/CTestScriptsUnitTests/TriBITS_GenericDriver.exe" "This test passed!  This test failed" "1"

--------------------------------------------------------------------------------


This test passed!  This test failed

End Result: TEST PASSED

--------------------------------------------------------------------------------

TEST_0: Return code = 1
TEST_0: Pass criteria = Match REGEX {This test passed} [PASSED]
TEST_0: Pass criteria = Not match REGEX {This test failed} [FAILED]
TEST_0: Pass criteria = WILL_FAIL (invert the above 'Pass critera') [PASSED]
TEST_0: Result = PASSED

================================================================================
```

Does this answer your questions?  If not, then we should get on Skype for Business or BlueJeans and discuss this to see what is missing.

**Comment by bkochuna** [comment]
Lets try getting on BlueJeans. I think our project has some differences so its a little hard to follow the instructions in your last post, but what I can follow does not give me the information I need.
